src/librustc_data_structures/bitvec.rs

```diff
@@ -254,15 +265,158 @@ impl BitMatrix {
     }
 }
 
+#[derive(Clone, Debug)]
+pub struct SparseBitMatrix<R, C>
+where
+    R: Idx,
+    C: Idx,
+{
+    vector: IndexVec<R, SparseBitSet<C>>,
+}
+
+impl<R: Idx, C: Idx> SparseBitMatrix<R, C> {
+    /// Create a new `rows x columns` matrix, initially empty.
+    pub fn new(rows: R, _columns: C) -> SparseBitMatrix<R, C> {
+        SparseBitMatrix {
+            vector: IndexVec::from_elem_n(SparseBitSet::new(), rows.index()),
+        }
+    }
+
+    /// Sets the cell at `(row, column)` to true. Put another way, insert
+    /// `column` to the bitset for `row`.
+    ///
+    /// Returns true if this changed the matrix, and false otherwise.
+    pub fn add(&mut self, row: R, column: C) -> bool {
+        self.vector[row].insert(column)
+    }
+
+    /// Do the bits from `row` contain `column`? Put another way, is
+    /// the matrix cell at `(row, column)` true?  Put yet another way,
+    /// if the matrix represents (transitive) reachability, can
+    /// `row` reach `column`?
+    pub fn contains(&self, row: R, column: C) -> bool {
+        self.vector[row].contains(column)
+    }
+
+    /// Returns those indices that are true in rows `a` and `b`.  This
+    /// is an O(n) operation where `n` is the number of elements
+    /// (somewhat independent from the actual size of the
+    /// intersection, in particular).
+    pub fn intersection(&self, a: R, b: R) -> Vec<C> {
+        let mut result = Vec::new();
+        let bit_set_a = &self.vector[a];
+        let bit_set_b = &self.vector[b];
+
+        for a_val in bit_set_a.iter().filter(|x| bit_set_b.contains(*x)) {
```

Comment:
I bet this method can just be removed. But if not, it could be made faster by adding a method `intersection` on `SparseBitset`.  I guess it would work something like this:

```rust
let mut result = SparseBitset::new();
for (key, value) in self.map {
    match other.map.get(key) {
        None => { /* other has no values with these key */ }
        Some(other_value) {
            if value & other_value != 0 {
                result.map.insert(key, value & other_value);
            }
        }
    }
}
```

This of course generates a new sparse-bitset.